fcrepo-api-x-binding/src/main/java/org/fcrepo/apix/binding/impl/RuntimeExtensionBinding.java

```diff
@@ -0,0 +1,131 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Simple extension binding based on runtime lookup and reasoning.
+ * <p>
+ * Polls the extension registry for all known extensions at time of transaction, and performs reasoning at runtime in
+ * order to bind a repository resource to the extensions that match it.
+ * </p>
+ */
+@Component(configurationPolicy = ConfigurationPolicy.REQUIRE)
+public class RuntimeExtensionBinding implements ExtensionBinding {
+
+    private static final Logger LOG = LoggerFactory.getLogger(RuntimeExtensionBinding.class);
+
+    private ExtensionRegistry extensionRegistry;
+
+    private OntologyService ontologySvc;
+
+    private Registry registry;
+
+    @Reference
+    public void setExtensionRegistry(ExtensionRegistry exr) {
+        extensionRegistry = exr;
+    }
+
+    @Reference
+    public void setOntologyService(OntologyService os) {
+        ontologySvc = os;
+    }
+
+    @Reference(target = "(org.fcrepo.apix.registry.role=default)")
+    public void setDelegateRegistry(Registry registry) {
+        this.registry = registry;
+    }
+
+    // (a) Determine the set of known extensions
+    // (b) for each extension, get its union ontology
+    // (c) for each union ontology, infer classes of the instance
+    // (d) Collect the list of classes
+    // (e) For each extension, see if its binding class is in that list. If so, collect it
+    // (f) return collected list of extensions.
+    @Override
+    public Collection<Extension> getExtensionsFor(WebResource resource) {
+
+        final Collection<Extension> extensions = extensionRegistry.getExtensions();
+        LOG.debug("(A) Got list of known extensions: {}", extensions);
+
+        final List<Extension> boundExtensions = new ArrayList<>();
+
+        for (final Extension e : extensions) {
+            LOG.debug("(B) Getting the ontology closure of extension {}", e.uri());
+            final Ontology o = ontologySvc.parseOntology(e.getResource());
+
+            LOG.debug("(C) Inferring classes over resource {} using ontologies from extension {}", resource.uri(), e
+                    .uri());
+            final Set<URI> inferredClasses = ontologySvc.inferClasses(resource.uri(), resource, o);
+
+            for (final URI inferredClass : inferredClasses) {
+                LOG.debug("(D) Found class {}", inferredClass);
+
+                for (final Extension candidate : extensions) {
+                    if (candidate.bindingClass().equals(inferredClass)) {
+                        boundExtensions.add(candidate);
+                        LOG.debug("(E) Class {} matches extension {}", candidate.uri(), candidate.bindingClass());
+                    }
+                }
+            }
+        }
+
+        return boundExtensions;
+
+        // This is perhaps a more concise way of performing the above^^, but the logging statements
+        // from the above are very useful.
+        //
```

Comment:
Ahh, you're already there... You can use the `peek()` method in the Stream API to handle logging.